stellargraph/core/element_data.py

```diff
@@ -119,59 +119,63 @@ class ElementData:
     than indexing pandas dataframes, series or indices.
 
     Args:
-        shared (dict of type name to pandas DataFrame): information for the elements of each type
+        shared (pandas DataFrame): information for each element
+        type_ranges (dict of type name to range): the ranges of each type of node; contiguous from 0, when sorted by the type names
     """
 
     # any columns that must be in the `shared` dataframes passed to `__init__` (this should be
     # overridden by subclasses as appropriate)
     _SHARED_REQUIRED_COLUMNS = []
 
-    def __init__(self, shared):
-        if not isinstance(shared, dict):
-            raise TypeError(f"shared: expected dict, found {type(shared)}")
+    def __init__(self, shared, type_ranges):
+        if not isinstance(shared, pd.DataFrame):
+            raise TypeError(
+                f"shared: expected pandas DataFrame, found {type(shared).__name__}"
+            )
 
-        for key, value in shared.items():
-            if not isinstance(value, pd.DataFrame):
-                raise TypeError(
-                    f"shared[{key!r}]: expected pandas DataFrame', found {type(value)}"
-                )
+        require_dataframe_has_columns("shared", shared, self._SHARED_REQUIRED_COLUMNS)
 
-            require_dataframe_has_columns(
-                f"features[{key!r}]", value, self._SHARED_REQUIRED_COLUMNS
+        if not isinstance(type_ranges, dict):
+            raise TypeError(
+                f"type_ranges: expected dict, found {type(type_ranges).__name__}"
             )
 
-        type_element_ilocs = {}
-        rows_so_far = 0
-        type_dfs = []
-
-        all_types = sorted(shared.keys())
-        type_sizes = []
+        sorted_ranges = sorted(type_ranges.items(), key=lambda x: x[0])
+        previous_end = 0
+        for type_name, type_range in sorted_ranges:
```

Comment:
Could `type_ranges` just be a sorted list of tuples `(range_start, type_name)` instead, and avoid having to do extra validation here, and just create the `range` objects here? It's just an internal type being passed around from `convert` right?

Reply:
Yeah, that's better; fewer things to get wrong. :+1: